File: visualbasic/hangman.vb
Imports System

REM  Copyright (c) 2021, Oliver Merkel.
REM  Please see the AUTHORS file for details.
REM  All rights reserved.

REM  Use of this code is governed by a
REM  MIT license that can be found in the LICENSE file.

Module hangman
  Dim maxFalseAttempts As Integer = 5

  Private Structure TResult
    Public result As String
    Public incorrect As Integer
  End Structure

  Private Sub Welcome()
    Console.WriteLine($"Guess my secret word with a maximum of {maxFalseAttempts} false attempts.")
  End Sub

  Private Function GetWord() As String
    Dim words() = IO.File.ReadAllLines("dict_en.txt")
    Dim word As String
    Dim r As New System.Random()
    word = ""
    While word.Equals("")
      word = words(r.Next(words.Length)).Trim()
    End While
    Return word
  End Function

  Private Function GetResult(word As String, guesses As String) As TResult
    Dim result As New TResult With {.result = "", .incorrect = 0}
    result.result = ""
    result.incorrect = 0
    For Each ch In word
      If guesses.Contains(ch.ToString().ToLower()) Then
        result.result = result.result + ch
      Else
        result.result = result.result + "_"
        result.incorrect = result.incorrect + 1
      End If
    Next
    Return result
  End Function

  Private Sub GameLoop()
    Welcome()
    Dim endOfGame = False
    Dim word = GetWord()
    Dim guesses = ""
    Dim falseAttempts = 0
    While Not endOfGame
      Dim test As New TResult With {.result = "", .incorrect = 0}
      test = GetResult(word, guesses)
      Console.WriteLine(test.result)
      If test.incorrect.Equals(0) Then
        endOfGame = True
        Console.WriteLine("You won!")
      Else
        Console.WriteLine("Guess a character:")
        Dim guess = Console.ReadLine().Trim().ToLower()
        If guess.Length <> 1 Then
          Console.WriteLine("Type a single character only! Try again...")
        ElseIf Not Char.IsLetter(guess, 0) Then
          Console.WriteLine("Type a alphabetic character! Try again...")
        ElseIf guesses.IndexOf(guess) <> -1 Then
          Console.WriteLine("You already guessed this character! Try again...")
        Else
          guesses = guesses + guess
          If Not word.ToLower().Contains(guess) Then
            falseAttempts = falseAttempts + 1
            Console.WriteLine("Wrong.")
            If falseAttempts = 1 Then
              Console.WriteLine("You have one false attempt.")
            Else
              Console.WriteLine($"You have {falseAttempts} false attempts.")
            End If
            If falseAttempts > maxFalseAttempts Then
              endOfGame = True
              Console.WriteLine("You lost!")
              Console.WriteLine("The secret word has been " + word.ToUpper())
            End If
          End If
        End If
      End If
    End While
  End Sub

  Private Function PlayAgain() As Boolean
    Dim yes = {"y", "yes", "sure"}
    Dim no = {"n", "no", "nope"}
    Dim isYes = False
    Dim isNo = False
    While Not isYes And Not isNo
      Console.WriteLine("Play again?")
      Dim answer = Console.ReadLine().Trim().ToLower()
      isYes = yes.Contains(answer)
      isNo = no.Contains(answer)
      If Not isYes And Not isNo Then
        Console.WriteLine("Sorry? Please answer with yes or no...")
      End If
    End While
    Return isYes
  End Function

  Public Sub Main(args As String())
    Dim keepPlaying = True
    While keepPlaying
      GameLoop()
      keepPlaying = PlayAgain()
    End While
    Console.WriteLine("Bye.")
  End Sub
End Module

File: fs/Hangman.fs
open System
open System.IO

(*
// Copyright (c) 2021, Oliver Merkel.
// Please see the AUTHORS file for details.
// All rights reserved.
//
// Use of this code is governed by a
// MIT license that can be found in the LICENSE file.
*)

type TestResult = { result: string; incorrect: int; }

let FILENAME = "dict_en.txt"
let MAXFALSEATTEMPTS = 5

let welcome() =
    Console.WriteLine("Guess my secret word with a maximum of " + MAXFALSEATTEMPTS.ToString() + " false attempts.")

let playAgain() =
    let yes = [ "y"; "yes"; "sure" ]
    let no = [ "n"; "no"; "nope" ]
    let mutable isYes = false
    let mutable isNo = false
    while not isYes && not isNo do
        Console.Write("Play again? ")
        let answer = (Console.ReadLine()).ToLower()
        isYes <- List.contains answer yes
        isNo <- List.contains answer no
        if not isYes && not isNo then
            Console.WriteLine("Sorry? Please answer with yes or no...")
    isYes

let getWord() =
    let words = File.ReadAllLines(FILENAME)
    let idx = (new System.Random()).Next(0, words.Length)
    words.[idx]

let getTestResult( word: string, guesses: string ) =
    let mutable result = String.map(fun c -> if guesses.Contains c then c else '_') word
    let mutable incorrect = (String.filter(fun c -> c = '_') result).Length
    { TestResult.result= result; TestResult.incorrect= incorrect; }

let gameLoop() =
    welcome()
    let mutable word = getWord()
    let mutable endOfGame = false
    let mutable falseAttempts = 0
    let mutable guesses = ""
    while not endOfGame do
        let test = getTestResult( word, guesses )
        Console.WriteLine(test.result)
        if test.incorrect = 0 then
            endOfGame <- true
            Console.WriteLine("You won!")
        else
            Console.WriteLine("Guess a character: ")
            let guess = (Console.ReadLine()).ToLower()
            if guess.Length <> 1 then
                Console.WriteLine("Type a single character only! Try again...")
            elif not (Char.IsLetter(guess, 0)) then
                Console.WriteLine("Type a alphabetic character! Try again...")
            elif guesses.Contains guess then
                Console.WriteLine("You already guessed this character! Try again...");
            else
                guesses <- guesses + guess
                if not (word.Contains guess) then
                    falseAttempts <- falseAttempts + 1
                    Console.WriteLine("Wrong.")
                    if falseAttempts = 1 then
                        Console.WriteLine("You have one false attempt.")
                    else
                        Console.WriteLine("You have " + falseAttempts.ToString() + " false attempts.")
                    if falseAttempts > MAXFALSEATTEMPTS then
                        endOfGame <- true
                        Console.WriteLine("You lost!")
                        Console.WriteLine("The secret word has been " + word.ToUpper())

let mainLoop() =
    let mutable keepPlaying = true
    while keepPlaying do
        gameLoop()
        keepPlaying <- playAgain()
    Console.WriteLine("Bye.")

[<EntryPoint>]
let main (_ : string[]) =
    mainLoop()
    0;;
